x-pack/plugins/alerting/server/invalidate_pending_api_keys/task.ts

```diff
@@ -189,37 +157,172 @@ function taskRunner(
   };
 }
 
-async function invalidateApiKeys(
-  logger: Logger,
-  savedObjectsClient: SavedObjectsClientContract,
-  apiKeysToInvalidate: SavedObjectsFindResponse<InvalidatePendingApiKey>,
-  encryptedSavedObjectsClient: EncryptedSavedObjectsClient,
-  securityPluginStart?: SecurityPluginStart
-) {
+interface ApiKeyIdAndSOId {
+  id: string;
+  apiKeyId: string;
+}
+
+interface RunInvalidateOpts {
+  config: AlertingConfig;
+  encryptedSavedObjectsClient: EncryptedSavedObjectsClient;
+  logger: Logger;
+  savedObjectsClient: SavedObjectsClientContract;
+  security?: SecurityPluginStart;
+}
+export async function runInvalidate({
+  config,
+  encryptedSavedObjectsClient,
+  logger,
+  savedObjectsClient,
+  security,
+}: RunInvalidateOpts) {
+  const configuredDelay = config.invalidateApiKeysTask.removalDelay;
+  const delay: string = timePeriodBeforeDate(new Date(), configuredDelay).toISOString();
+
+  let hasApiKeysPendingInvalidation = true;
   let totalInvalidated = 0;
+  const excludedSOIds = new Set<string>();
+  do {
+    const filter = getFindFilter(delay, [...excludedSOIds]);
+    const apiKeysToInvalidate = await savedObjectsClient.find<InvalidatePendingApiKey>({
+      type: API_KEY_PENDING_INVALIDATION_TYPE,
+      filter,
+      page: 1,
+      sortField: 'createdAt',
+      sortOrder: 'asc',
+      perPage: PAGE_SIZE,
+    });
+
+    if (apiKeysToInvalidate.total > 0) {
+      const { apiKeyIdsToExclude, apiKeyIdsToInvalidate } = await getApiKeyIdsToInvalidate({
```

Comment:
Yea it took me a bit to work out the logic as well. I thought there was a bug with the paging! But the intent is that we're deleting it from the SO index so if you have 110 keys to invalidate, you get the first 100 and delete them, then the next iteration, you'd get the remaining 10. That's why we need to add the exclusion ID filter because we are not deleting those so we need to exclude them

WRT to the refresh, I believe Kibana system indices are fast refresh indices so we shouldn't need to force a refresh? Does that sound right?

Reply:
@pmuellr WDYT of removing the `do...while` loop altogether and just having each run of the task query for and invalidate up to 100? The task is scheduled to run every 5 minutes anyway and that would simplify some of this logic.

Reply:
I added some comments to try to explain in https://github.com/elastic/kibana/commit/cf85e78edcb813d46da2c53922ad9478df597111

Reply:
Asked in slack and because they are fast refresh indices and the default refresh option is `wait_for`, we shouldn't need to perform any explicit refreshes.

Reply:
> WDYT of removing the do...while loop altogether and just having each run of the task query for and invalidate up to 100?

I was thinking about something like that yesterday.  The potential problem would be if we can't catch up.  Hard to imagine that happening though.  Also, I do remember "debugging" these API keys with someone once, who was "watching" them get invalidated - I had to explain the delay, etc.  So the delay could be longer than usual. 

So ... seems fine to me to make a change like that.  

Perhaps we should add a debug log providing numbers: total # of SO's, total invalidated this time, kept valid because of ad hoc runs, etc, so we'll get some kinda-easy visibility if we think it's having problems.

Reply:
I'm going to make an issue for this for followup.

Reply:
https://github.com/elastic/kibana/issues/181198